File: apps/ios/Sources/Core/MessageActions/MessageViewport.swift
import Foundation

enum ViewportMove: Equatable, Sendable { case none, latest, restore(ScrollAnchor) }
/// History insertion preserves visible offset. Only committed new sync IDs count as incoming.
struct MessageViewport {
    private var initialized = false
    private var followsLatest = false
    private var visible: ScrollAnchor?
    private var incoming: Set<String> = []
    var incomingCount: Int { incoming.count }
    mutating func reset() { initialized = false; followsLatest = false; visible = nil; incoming = [] }
    mutating func initialize(restored: ScrollAnchor?) -> ViewportMove {
        guard !initialized else { return .none }
        initialized = true; visible = restored; followsLatest = restored == nil
        return restored.map(ViewportMove.restore) ?? .latest
    }
    mutating func observed(anchor: ScrollAnchor?, atLatest: Bool) {
        visible = anchor; followsLatest = atLatest; if atLatest { incoming = [] }
    }
    func olderPageCommitted() -> ViewportMove { visible.map(ViewportMove.restore) ?? .none }
    mutating func incomingCommitted(_ ids: Set<String>) throws -> ViewportMove {
        guard ids.allSatisfy(actionID) else { throw MessageActionError.invalidResponse }
        guard initialized, !ids.isEmpty else { return .none }
        if followsLatest { return .latest }
        incoming.formUnion(ids); return .none
    }
    mutating func showLatest() -> ViewportMove { followsLatest = true; incoming = []; return .latest }
    mutating func deleted(_ messageId: String) { incoming.remove(messageId); if visible?.messageId == messageId { visible = nil } }
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/messageactions/MessageViewport.kt
package chat.rogi.rogichat.core.messageactions

sealed interface ViewportMove {
    data object None : ViewportMove
    data object Latest : ViewportMove
    data class Restore(val anchor: ScrollAnchor) : ViewportMove
}
/** Feed only IDs newly committed by sync; history pages are not incoming messages. */
class MessageViewport {
    private var initialized = false
    private var followsLatest = false
    private var visible: ScrollAnchor? = null
    private val incoming = mutableSetOf<String>()
    val incomingCount get() = incoming.size
    fun reset() { initialized = false; followsLatest = false; visible = null; incoming.clear() }
    fun initialize(restored: ScrollAnchor?): ViewportMove {
        if (initialized) return ViewportMove.None
        initialized = true; visible = restored; followsLatest = restored == null
        return restored?.let(ViewportMove::Restore) ?: ViewportMove.Latest
    }
    fun observed(anchor: ScrollAnchor?, atLatest: Boolean) {
        visible = anchor; followsLatest = atLatest; if (atLatest) incoming.clear()
    }
    fun olderPageCommitted(): ViewportMove = visible?.let(ViewportMove::Restore) ?: ViewportMove.None
    fun incomingCommitted(ids: Set<String>): ViewportMove {
        ids.forEach(::actionId)
        if (!initialized || ids.isEmpty()) return ViewportMove.None
        if (followsLatest) return ViewportMove.Latest
        incoming.addAll(ids); return ViewportMove.None
    }
    fun showLatest(): ViewportMove { followsLatest = true; incoming.clear(); return ViewportMove.Latest }
    fun deleted(messageId: String) { incoming.remove(messageId); if (visible?.messageId == messageId) visible = null }
}
